sidecar/src/service/agent_info.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+use crate::one_way_shared_memory::{open_named_shm, OneWayShmReader, OneWayShmWriter};
+use crate::primary_sidecar_identifier;
+use base64::prelude::BASE64_URL_SAFE_NO_PAD;
+use base64::Engine;
+use data_pipeline::agent_info::schema::AgentInfoStruct;
+use data_pipeline::agent_info::{fetch_info_with_state, FetchInfoStatus};
+use datadog_ipc::platform::NamedShmHandle;
+use ddcommon::Endpoint;
+use futures::future::Shared;
+use futures::FutureExt;
+use http::uri::PathAndQuery;
+use manual_future::ManualFuture;
+use std::ffi::CString;
+use std::hash::{Hash, Hasher};
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+use tokio::time::sleep;
+use tracing::{error, warn};
+use zwohash::{HashMap, ZwoHasher};
+
+#[derive(Default, Clone)]
+pub struct AgentInfos(Arc<Mutex<HashMap<Endpoint, AgentInfoFetcher>>>);
+
+impl AgentInfos {
+    pub fn query_for(&self, endpoint: Endpoint) -> AgentInfoGuard {
+        let mut infos_guard = self.0.lock().unwrap();
+        if let Some(info) = infos_guard.get_mut(&endpoint) {
+            info.rc += 1;
+        } else {
+            infos_guard.insert(
+                endpoint.clone(),
+                AgentInfoFetcher::new(self.clone(), endpoint.clone()),
+            );
+        }
+
+        AgentInfoGuard {
+            infos: self.clone(),
+            endpoint,
+        }
+    }
+}
+
+pub struct AgentInfoGuard {
+    infos: AgentInfos,
+    endpoint: Endpoint,
+}
+
+impl AgentInfoGuard {
+    pub fn get(&self) -> Shared<ManualFuture<AgentInfoStruct>> {
+        let infos_guard = self.infos.0.lock().unwrap();
+        let infos = infos_guard.get(&self.endpoint).unwrap();
+        infos.infos.clone()
+    }
+}
+
+impl Drop for AgentInfoGuard {
+    fn drop(&mut self) {
+        let mut infos_guard = self.infos.0.lock().unwrap();
+        let info = infos_guard.get_mut(&self.endpoint).unwrap();
+        info.last_update = Instant::now();
+        info.rc -= 1;
+    }
+}
+
+pub struct AgentInfoFetcher {
+    last_update: Instant,
+    rc: u32,
+    infos: Shared<ManualFuture<AgentInfoStruct>>,
+}
+
+impl AgentInfoFetcher {
```

Comment:
This file has no comments or tests 😞 I also find it confusing that this has the same name `AgentInfoFetcher` as the `AgentInfoFetcher` in `data-pipeline`. Yes, I know that there are differences, but couldn't the `break` check, and the _processing_ be moved behind a `trait` and there only be one impl of the loop?

Reply:
I wasn't intentionally choosing the same name. Maybe I can call it AgentInfoWriter, but not sure if that helps?

You theoretically could do it, but the actual amount of shared code would be quite low & the boilerplate for the sharing of code quite large in comparison? I feel like fetch_info_with_state is the right level of abstraction for this.

I can though add at least a test for the happy path here.

Reply:
Fine, let's not try to share more code then.

Please add a happy path test and some comments. I mean at least something like:

```rust
//! This file contains code for fetching and sharing the info from the Datadog Agent
//! ... etc
```

Reply:
Yes, I will do, just haven't gotten round to it yet.

Reply:
@bantonsson Done :-)